### PR TITLE
pid1: fix draw_cylon()

### LINKED ISSUE #30121: Possible stack-buffer-overflow in PID 1
I somehow managed to trigger this while testing #30119:

```
Nov 21 13:36:07 H systemd[1]: Reached target nss-lookup.target.
Nov 21 13:36:07 H systemd[1]: sysinit.target: starting held back, waiting for: systemd-hwdb-update.service
Nov 21 13:36:07 H systemd[1]: Event source 0x60f000000400 (mount-monitor-dispatch) left rate limit state.
Nov 21 13:36:34 H systemd[1]: =================================================================
Nov 21 13:36:34 H systemd[1]: ==1==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7f06b1a16c66 at pc 0x7f06b62a8a75 bp 0x7ffdbe4696d0 sp 0x7ffdbe468e90
Nov 21 13:36:34 H systemd[1]: READ of size 49 at 0x7f06b1a16c66 thread T0 (systemd)
Nov 21 13:36:34 H systemd[1]:     #0 0x7f06b62a8a74  (/usr/lib/clang/16/lib/linux/libclang_rt.asan-x86_64.so+0xa8a74) (BuildId: 6d967465ddfb168365d0ee0296553c15bd2c72de)
Nov 21 13:36:34 H systemd[1]:     #1 0x7f06b5c83422 in status_vprintf /systemd-meson-build/../build/src/core/show-status.c:90:38
Nov 21 13:36:34 H systemd[1]:     #2 0x7f06b5b8de88 in manager_status_printf /systemd-meson-build/../build/src/core/manager.c:4515:9
Nov 21 13:36:34 H systemd[1]:     #3 0x7f06b5b93c55 in manager_print_jobs_in_progress /systemd-meson-build/../build/src/core/manager.c:291:17
Nov 21 13:36:34 H systemd[1]:     #4 0x7f06b5b93c55 in manager_dispatch_jobs_in_progress /systemd-meson-build/../build/src/core/manager.c:3260:9
Nov 21 13:36:34 H systemd[1]:     #5 0x7f06b4e89d65 in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4195:21
Nov 21 13:36:34 H systemd[1]:     #6 0x7f06b4e88a3d in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4808:21
Nov 21 13:36:34 H systemd[1]:     #7 0x7f06b4e8c94f in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4869:21
Nov 21 13:36:34 H systemd[1]:     #8 0x7f06b5b7fd48 in manager_loop /systemd-meson-build/../build/src/core/manager.c:3324:21
Nov 21 13:36:34 H systemd[1]:     #9 0x55ea829854a2 in invoke_main_loop /systemd-meson-build/../build/src/core/main.c:1961:33
Nov 21 13:36:34 H systemd[1]:     #10 0x55ea82979ddf in main /systemd-meson-build/../build/src/core/main.c:3126:13
Nov 21 13:36:34 H systemd[1]:     #11 0x7f06b4045ccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
Nov 21 13:36:34 H systemd[1]:     #12 0x7f06b4045d89 in __libc_start_main (/usr/lib/libc.so.6+0x27d89) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
Nov 21 13:36:34 H systemd[1]:     #13 0x55ea829752b4 in _start (/usr/lib/systemd/systemd+0x182b4) (BuildId: d650fd741b398714d79992e5668c84b7e864132c)
Nov 21 13:36:34 H systemd[1]: 
Nov 21 13:36:34 H systemd[1]: Address 0x7f06b1a16c66 is located in stack of thread T0 (systemd)
Nov 21 13:36:34 H systemd[1]:  at offset 102 in frame
Nov 21 13:36:34 H systemd[1]:     #0 0x7f06b5b92f5f in manager_dispatch_jobs_in_progress /systemd-meson-build/../build/src/core/manager.c:3254
Nov 21 13:36:34 H systemd[1]: 
Nov 21 13:36:34 H systemd[1]:   This frame has 7 object(s):
Nov 21 13:36:34 H systemd[1]:     [32, 40) 'j.i' (line 234)
Nov 21 13:36:34 H systemd[1]:     [64, 102) 'cylon.i' (line 236)
Nov 21 13:36:34 H systemd[1]:     [144, 152) 'timeout.i' (line 238) <== Memory access at offset 102 partially underflows this variable
Nov 21 13:36:34 H systemd[1]:     [176, 192) '__unique_prefix_i12.i' (line 247)
Nov 21 13:36:34 H systemd[1]:     [208, 237) 'job_of_n.i' (line 263)
Nov 21 13:36:34 H systemd[1]:     [272, 336) '.compoundliteral.i' (line 308)
Nov 21 13:36:34 H systemd[1]:     [368, 432) '.compoundliteral168.i' (line 308)
Nov 21 13:36:34 H systemd[1]: HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
Nov 21 13:36:34 H systemd[1]:       (longjmp and C++ exceptions *are* supported)
Nov 21 13:36:34 H systemd[1]: SUMMARY: AddressSanitizer: stack-buffer-overflow (/usr/lib/clang/16/lib/linux/libclang_rt.asan-x86_64.so+0xa8a74) (BuildId: 6d967465ddfb168365d0ee0296553c15bd2c72de)
Nov 21 13:36:34 H systemd[1]: Shadow bytes around the buggy address:
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16980: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16a00: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16a80: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16b00: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16b80: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Nov 21 13:36:34 H systemd[1]: =>0x7f06b1a16c00: f1 f1 f1 f1 00 f2 f2 f2 00 00 00 00[06]f2 f2 f2
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16c80: f2 f2 00 f2 f2 f2 f8 f8 f2 f2 00 00 00 05 f2 f2
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16d00: f2 f2 00 00 00 00 00 00 00 00 f2 f2 f2 f2 00 00
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16d80: 00 00 00 00 00 00 f3 f3 f3 f3 f3 f3 00 00 00 00
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16e00: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Nov 21 13:36:34 H systemd[1]:   0x7f06b1a16e80: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Nov 21 13:36:34 H systemd[1]: Shadow byte legend (one shadow byte represents 8 application bytes):
Nov 21 13:36:34 H systemd[1]:   Addressable:           00
Nov 21 13:36:34 H systemd[1]:   Partially addressable: 01 02 03 04 05 06 07
Nov 21 13:36:34 H systemd[1]:   Heap left redzone:       fa
Nov 21 13:36:34 H systemd[1]:   Freed heap region:       fd
Nov 21 13:36:34 H systemd[1]:   Stack left redzone:      f1
Nov 21 13:36:34 H systemd[1]:   Stack mid redzone:       f2
Nov 21 13:36:34 H systemd[1]:   Stack right redzone:     f3
Nov 21 13:36:34 H systemd[1]:   Stack after return:      f5
Nov 21 13:36:34 H systemd[1]:   Stack use after scope:   f8
Nov 21 13:36:34 H systemd[1]:   Global redzone:          f9
Nov 21 13:36:34 H systemd[1]:   Global init order:       f6
Nov 21 13:36:34 H systemd[1]:   Poisoned by user:        f7
Nov 21 13:36:34 H systemd[1]:   Container overflow:      fc
Nov 21 13:36:34 H systemd[1]:   Array cookie:            ac
Nov 21 13:36:34 H systemd[1]:   Intra object redzone:    bb
Nov 21 13:36:34 H systemd[1]:   ASan internal:           fe
Nov 21 13:36:34 H systemd[1]:   Left alloca redzone:     ca
Nov 21 13:36:34 H systemd[1]:   Right alloca redzone:    cb
Nov 21 13:36:34 H systemd[1]: 
Nov 21 13:36:34 H systemd[1]: Command:
Nov 21 13:36:34 H systemd[1]: /usr/lib/systemd/systemd
Nov 21 13:36:34 H systemd[1]: --
Nov 21 13:36:34 H systemd[1]: systemd.unit=testsuite.target
Nov 21 13:36:34 H systemd[1]: systemd.wants=testsuite-01.service
Nov 21 13:36:34 H systemd[1]: systemd.wants=end.service
Nov 21 13:36:34 H systemd[1]: foo
Nov 21 13:36:34 H systemd[1]: -z
Nov 21 13:36:34 H systemd[1]: bar
Nov 21 13:36:34 H systemd[1]: ---
Nov 21 13:36:34 H systemd[1]: baz
Nov 21 13:36:34 H systemd[1]: 
Nov 21 13:36:34 H systemd[1]: 
Nov 21 13:36:34 H systemd[1]: ==1==ABORTING
```

Full log: https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/7932/artifact//systemd-centos-ci/artifacts_all/artifacts___hfxqzn/vagrant-logs.q5h/vagrant-arch-sanitizers-clang-testsuite.fz6/TEST-01-BASIC_sanitizers-nspawn_FAIL.log

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #30121.

### CODE REVIEW

File: src/core/manager.c [hunk 1]
```diff
@@ -214,8 +214,10 @@ static void draw_cylon(char buffer[], size_t buflen, unsigned width, unsigned po
                 if (pos < width-1)
                         p = mempset(p, ' ', width-1-pos);
                 if (log_get_show_color())
-                        strcpy(p, ANSI_NORMAL);
+                        stpcpy(p, ANSI_NORMAL);
```
Comment (bluca) [inline src/core/manager.c]: did you mean `p = stpcpy()`?

Comment (yuwata) [reply src/core/manager.c]: Ouch!

File: src/systemd/id_full
Comment (bluca) [inline src/systemd/id_full:1]: git add by mistake?

Comment (yuwata) [reply src/systemd/id_full:1]: I have no idea when such file is created... Anyway, dropped.

**Review by bluca — APPROVED** [review]

File: src/shared/pretty-print.h [hunk 2]
```diff
@@ -4,6 +4,10 @@
 #include "glyph-util.h"
 #include "terminal-util.h"
 
+#define CYLON_BUFFER_EXTRA (2*STRLEN(ANSI_RED) + STRLEN(ANSI_HIGHLIGHT_RED) + 2*STRLEN(ANSI_NORMAL))
+
+void draw_cylon(char buffer[], size_t buflen, unsigned width, unsigned pos);
```
Comment (poettering) [inline src/shared/pretty-print.h:9]: maybe change first arg to `char buffer[static CYLON_BUFFER_EXTRA]`

Comment (yuwata) [reply src/shared/pretty-print.h:9]: No, the buffer size must be larger than `CYLON_BUFFER_EXTRA`.

**Pull request merged**